### PR TITLE
Adding sections on checks and suppressing checks

### PR DESCRIPTION
**Pull request by gusthoff**

### CODE REVIEW

File: content/courses/advanced-ada/chapters/exceptions.rst [hunk 1]
```diff
@@ -235,6 +235,589 @@ the call to :ada:`Assert` is not ignored.
...
+       A2 := T2_Class (Copy (A3));
+
+    end Show_Tag_Check;
+
+Here, the tag of the object returned by the :ada:`Copy` function is
+:ada:`T1'Tag`. However, we cannot convert an object with this tag to an
+object of :ada:`T2_Class`, so the tag check fails.
+
+
+Accessibility Check
```
Comment (pat-rogers) [inline content/courses/advanced-ada/chapters/exceptions.rst:590]: IMHO this one is too complex for an intro course.

File: content/courses/advanced-ada/chapters/exceptions.rst [hunk 2]
```diff
@@ -235,6 +235,589 @@ the call to :ada:`Assert` is not ignored.
...
+    begin
+       P (T1_Class (A1));
+
+    end Show_Accessibility_Check;
+
+Again, the check fails in the :ada:`T1_Class (A1)` conversion and raises a
+:ada:`Program_Error` exception.
+
+
+Allocation Check
```
Comment (pat-rogers) [inline content/courses/advanced-ada/chapters/exceptions.rst:668]: IMHO this one is too complex for an intro course.

File: content/courses/advanced-ada/chapters/exceptions.rst [hunk 3]
```diff
@@ -235,6 +235,589 @@ the call to :ada:`Assert` is not ignored.
...
+| :ada:`Tag_Check`            | :ada:`Constraint_Error` |
++-----------------------------+-------------------------+
+| :ada:`Accessibility_Check`  | :ada:`Program_Error`    |
++-----------------------------+-------------------------+
+| :ada:`Allocation_Check`     | :ada:`Program_Error`    |
++-----------------------------+-------------------------+
+| :ada:`Elaboration_Check`    | :ada:`Program_Error`    |
++-----------------------------+-------------------------+
+| :ada:`Storage_Check`        | :ada:`Storage_Error`    |
++-----------------------------+-------------------------+
```
[the remark this answers is not in the document]
Comment (gusthoff) [reply content/courses/advanced-ada/chapters/exceptions.rst:269]: Not for the moment, as I'm avoiding Ada 2022 features. However, thanks for the hint! I'll add a todo item.

File: content/courses/advanced-ada/chapters/exceptions.rst [hunk 4]
```diff
@@ -235,6 +235,589 @@ the call to :ada:`Assert` is not ignored.
...
+    begin
+       R := (Valid  => True,
+             Counter => 10);
+    end Show_Discriminant_Check;
+
+Here, :ada:`R`\ 's discriminant (:ada:`Valid`) is :ada:`False`, so we cannot
+assign an object whose :ada:`Valid` discriminant is :ada:`True`.
+
+Also, when accessing a component, the discriminant check ensures that this
+component exist (for the current discriminant values):
```
[the remark this answers is not in the document]
Comment (gusthoff) [reply content/courses/advanced-ada/chapters/exceptions.rst:359]: Sounds good, I'll change it.

File: content/courses/advanced-ada/chapters/exceptions.rst [hunk 5]
```diff
@@ -235,6 +235,589 @@ the call to :ada:`Assert` is not ignored.
...
+       I := Arr (11);
+    end Show_Range_Check;
+
+In this example, the range of the input array :ada:`Arr` |mdash| passed to
+parameter :ada:`A` of the :ada:`Value_Of` function |mdash| is 1 to 10. In the
+call to :ada:`Value_Of`, we're trying to access position 11, which is outside
+this range. Therefore, the range check fails.
+
+
+Tag Check
```
[the remark this answers is not in the document]
Comment (gusthoff) [reply content/courses/advanced-ada/chapters/exceptions.rst:545]: This is an advanced course. :-)

File: content/courses/advanced-ada/chapters/exceptions.rst (hunk 1 above)
Comment (gusthoff) [reply content/courses/advanced-ada/chapters/exceptions.rst:590]: Same here: this is an advanced course, so I guess it should be OK.

File: content/courses/advanced-ada/chapters/exceptions.rst [hunk 6]
```diff
@@ -235,6 +235,589 @@ the call to :ada:`Assert` is not ignored.
...
+       I   : Integer;
+
+    begin
+       I := Arr (11);
+    end Show_Range_Check;
+
+In this example, the range of the input array :ada:`Arr` |mdash| passed to
+parameter :ada:`A` of the :ada:`Value_Of` function |mdash| is 1 to 10. In the
+call to :ada:`Value_Of`, we're trying to access position 11, which is outside
+this range. Therefore, the range check fails.
```
[the remark this answers is not in the document]
Comment (gusthoff) [reply content/courses/advanced-ada/chapters/exceptions.rst:542]: Interesting. For the example that I used, the application exits with this exception:

```raised CONSTRAINT_ERROR : show_range_check.adb:14 range check failed```

Is the error message (inserted by GNAT) wrong?

**Comment by gusthoff** [comment]
Thanks for the review. I changed some parts according to your suggestions.

Regarding the complexity of some code examples, I think this should be OK, as this is an Advanced Ada course. We'll have more of those complex examples in upcoming chapters.

File: content/courses/advanced-ada/chapters/exceptions.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (gusthoff) [reply content/courses/advanced-ada/chapters/exceptions.rst:547]: Nevermind, (I think) I'm seeing my mistake now.

[the remark this answers is not in the document]
Comment (gusthoff) [reply content/courses/advanced-ada/chapters/exceptions.rst]: I changed the text, but there's still the question whether `Arr (11)` is an index check or a range check. My interpretation, based on the exception message, is that the compiler inserts a range check here to verify whether `11` is in the `1 .. 10` range.

[the remark this answers is not in the document]
Comment (gusthoff) [reply content/courses/advanced-ada/chapters/exceptions.rst]: (I do agree with you, though, that this _looks like_ an index check. I was actually surprised by the exception message...)

[the remark this answers is not in the document]
Comment (gusthoff) [reply content/courses/advanced-ada/chapters/exceptions.rst]: I simplified the example a little bit, so it's hopefully not ambiguous anymore.

**Comment by pat-rogers** [comment]
> This is an advanced course. :-)

Oops! Never mind... :-)

**Review by pat-rogers — APPROVED** [review]

**Pull request merged**